### PR TITLE
Update DB on Job Complete

### PR DESCRIPTION
**Pull request by btylerburton**
- adds new single record dcat source;
- updates ckan import for better mocking;
- sets error status on internal record prior to raising error; 
- creates job status update dict for report method;
- updates job status on job complete;
- updates record status on success;
- adds tests;
- cleans up mocking;

# Pull Request

Related to https://github.com/GSA/data.gov/issues/4762

## About

Updates db after job complete.

## PR TASKS

- [X] The actual code changes.
- [X] Tests written and passed.
- [ ] Any changes to docs?
- [ ] Bumped version number in [pyproject.toml](https://github.com/GSA/datagov-harvesting-logic/blob/8078c5b9f015ca7fb8a142e4e4b2e22ad2fd49b1/pyproject.toml#L3) (also checked on [PyPi](https://pypi.org/project/ckanext-geodatagov/#history)).

### CODE REVIEW

**Comment by btylerburton** [comment]
FYI There is a test case where the job experiences a CriticalException, and should be marked as error. This is not covered. I will either add to this PR or make a new one on Monday taking that into account.

**Comment by btylerburton** [comment]
@GSA/data-gov-dev-team PR is updated and ready to go 🐑

File: tests/integration/full-flow/test_harvest_full_flow.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        organization_data,
+        source_data_dcatus_single_record,
+        single_internal_record,
+    ):
+        CKANMOCK.return_value.action = "ok"
+        interface.add_organization(organization_data)
+        interface.add_harvest_source(source_data_dcatus_single_record)
+        harvest_job = interface.add_harvest_job(
+            {
+                "status": "pending",
```
Comment (rshewitt) [inline tests/integration/full-flow/test_harvest_full_flow.py]: "pending" is no longer a value used in the enum for harvest job status. it should either be "new" or "manual"

Comment (btylerburton) [reply tests/integration/full-flow/test_harvest_full_flow.py]: Did you change that in your PR? Currently still valid here: https://github.com/GSA/datagov-harvesting-logic/blob/main/database/models.py#L53-L54

**Comment by rshewitt** [comment]
merge conflicts need to be resolved after adding the model updates

File: database/interface.py [hunk 2]
```diff
@@ -211,37 +216,46 @@ def add_harvest_error(self, error_data: dict, error_type: str):
             self.db.rollback()
             return None
 
-    def get_all_errors_of_job(self, job_id: str) -> [[dict], [dict]]:
-        job = self.get_harvest_job(job_id)
-        job_errors = list(map(HarvesterDBInterface._to_dict, job.errors))
+    def add_harvest_record_error(self, error_data: dict):
```
Comment (rshewitt) [inline database/interface.py:219]: despite this function being almost identical to `add_harvest_job_error` they represent different things. i get the reason to do it.

File: database/interface.py [hunk 3]
```diff
@@ -211,37 +216,46 @@ def add_harvest_error(self, error_data: dict, error_type: str):
...
+            return list(map(HarvesterDBInterface._to_dict, job.errors))
+        else:
+            return []
 
-        return [job_errors, record_errors]
+    def get_harvest_record_errors_by_job(self, job_id: str):
+        job = self.get_harvest_job(job_id)
+        if job:
+            return [error for record in job.records for error in record.errors]
+        else:
```
Comment (rshewitt) [inline database/interface.py:242]: easier to read than a comment on and use of itertools.chain

File: database/interface.py [hunk 4]
```diff
@@ -211,37 +216,46 @@ def add_harvest_error(self, error_data: dict, error_type: str):
...
+            return []
 
-    def get_harvest_job_error(self, job_id: str) -> dict:
-        result = self.db.query(HarvestJobError).filter_by(harvest_job_id=job_id).first()
+    def get_harvest_error(self, error_id: str) -> dict:
+        job_query = self.db.query(HarvestJobError).filter_by(id=error_id)
+        record_query = self.db.query(HarvestRecordError).filter_by(id=error_id)
+        result = job_query.union(record_query).first()
```
Comment (rshewitt) [inline database/interface.py:248]: we talked about whether we would need this. let's decide later.

File: database/models.py [hunk 5]
```diff
@@ -10,6 +10,7 @@ class Base(DeclarativeBase):
     id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
 
 
+# For ref: https://stackoverflow.com/questions/22698478/what-is-the-difference-between-the-declarative-base-and-db-model
```
Comment (rshewitt) [inline database/models.py:13]: would be worthwhile to revisit this. is what we're currently doing not already handled by `flask-sqlalchemy`? no point in duplicating functionality

**Review by rshewitt — APPROVED** [review]
looks good!

**Pull request merged**